rosetta/retriever/retriever_test.go

```diff
@@ -0,0 +1,1196 @@
+			},
+		}
+
+		ret := &Retriever{index: mock}
+		_, _, err := ret.Oldest()
+
+		assert.Error(t, err)
+	})
+}
+
+func TestRetriever_Current(t *testing.T) {
+	testHeight := uint64(42)
+	testTime := time.Time{} // 1/1/1970
+	testBlockID := identifier.Block{
+		Index: 42,
+		Hash:  "499b933f5ecd062d5ff7914218a40f8abf0efee9267d46ae827c938f8a5c18ae",
+	}
+	testHeader := &flow.Header{
+		Height:    testHeight,
+		Timestamp: testTime,
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
+		mock := &mocks.Reader{
+			LastFunc: func() (uint64, error) {
+				return testHeight, nil
+			},
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+		}
+
+		ret := &Retriever{index: mock}
+		blockID, blockTime, err := ret.Current()
+
+		if assert.NoError(t, err) {
+			assert.Equal(t, testTime, blockTime)
+			assert.Equal(t, testBlockID, blockID)
+		}
+	})
+
+	t.Run("handles index.Last failure", func(t *testing.T) {
+		mock := &mocks.Reader{
+			LastFunc: func() (uint64, error) {
+				return 0, mocks.DummyError
+			},
+		}
+
+		ret := &Retriever{index: mock}
+		_, _, err := ret.Current()
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles index.Header failure", func(t *testing.T) {
+		mock := &mocks.Reader{
+			LastFunc: func() (uint64, error) {
+				return testHeight, nil
+			},
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				return nil, mocks.DummyError
+			},
+		}
+
+		ret := &Retriever{index: mock}
+		_, _, err := ret.Current()
+
+		assert.Error(t, err)
+	})
+}
+
+func TestRetriever_Balances(t *testing.T) {
+	testBlockID := identifier.Block{
+		Index: 42,
+		Hash:  "2c5efefc2fafa000a3102f2931598d2d",
+	}
+	testAccount := identifier.Account{Address: "test"}
+	testCurrency1 := identifier.Currency{Symbol: "TEST1"}
+	testCurrency2 := identifier.Currency{Symbol: "TEST2"}
+	testCurrency3 := identifier.Currency{Symbol: "TEST3"}
+	testCurrencies := []identifier.Currency{testCurrency1, testCurrency2, testCurrency3}
+	testAmounts := []object.Amount{
+		{
+			Value:    "42",
+			Currency: testCurrency1,
+		},
+		{
+			Value:    "42",
+			Currency: testCurrency2,
+		},
+		{
+			Value:    "42",
+			Currency: testCurrency3,
+		},
+	}
+	testValue, err := cadence.NewValue(uint64(42))
+	require.NoError(t, err)
+
+	t.Run("nominal case", func(t *testing.T) {
+		validator := &mocks.Validator{
+			AccountFunc:  func(address identifier.Account) error { return nil },
+			BlockFunc:    func(block identifier.Block) (identifier.Block, error) { return block, nil },
+			CurrencyFunc: func(currency identifier.Currency) (identifier.Currency, error) { return currency, nil },
+		}
+		generator := &mocks.Generator{
+			GetBalanceFunc: func(symbol string) ([]byte, error) {
+				return []byte(`testScript`), nil
+			},
+		}
+		invoker := &mocks.Invoker{
+			ScriptFunc: func(height uint64, script []byte, parameters []cadence.Value) (cadence.Value, error) {
+				return testValue, nil
+			},
```

Comment:
I imagine we could add some asserts here to check it's called with the correct values? Same maybe for some of the other functions?